python/cudf/setup.py

```diff
@@ -101,3 +104,23 @@
     install_requires=install_requires,
     zip_safe=False,
 )
+
+
+protoc = None
+if "PROTOC" in os.environ and os.path.exists(os.environ["PROTOC"]):
+    protoc = os.environ["PROTOC"]
+else:
+    protoc = find_executable("protoc")
+if protoc is None:
+    sys.stderr.write("protoc not found")
+    sys.exit(1)
+
+
+for source in ["cudf/utils/metadata/orc_column_statistics.proto"]:
+    output = source.replace(".proto", "_pb2.py")
+
+    if not os.path.exists(output) or (
+        os.path.getmtime(source) > os.path.getmtime(output)
+    ):
+        sys.stderr.write("compiling " + source + "\n")
+        subprocess.check_call([protoc, "--python_out=.", source])
```

Comment:
I think this will run at every invocation of `setup.py` as opposed to just when a user is asking to build things. I.E. if someone uses `python setup.py help` it would still invoke this code. I think an approach to this would be to get the cmdclass returned from `versioneer.get_cmdclass()`, say as `tmp_cmdclass`, then create a new class that inherits from `tmp_cmdclass['build_py']`. Our new class could do `tmp_cmdclass['build_py'].run(self)` and then run our own code for building the protobuf into a Python file.

I'm not sure if this is a good approach or not though. Maybe @jakirkham or @jacobtomlinson have some ideas here?

Reply:
Would a user invoke this `setup.py` for something other than building? I had started implementing what you describe and can implement that - I'm just not sure whether to extend the `setup.py build_py` command or something else. From looking at other build scripts [1, 2], it seems like only `build_ext` (which I think is from Cython) and `install` are invoked and I don't think these indirectly invoke `build_py` though I could be wrong about that.

[1] https://github.com/trxcllnt/rapids-compose/blob/master/etc/bash-utils.sh#L1028
[2] https://github.com/rapidsai/cudf/blob/branch-0.16/build.sh#L179-L189

Reply:
There's `setup.py develop`, `setup.py help`, `setup.py bdist`, `setup.py sdist`, etc. `rapids-compose` likely doesn't cover all of these. My main concern is `setup.py help` isn't uncommon and we don't want the protobuf getting compiled if someone runs `setup.py help`.

Reply:
Just pushed with `build_ext` extended to do protobuf compilation. Took me some time bc I was relying on console output to verify the change was working instead of just doing a `clean-cudf-python && build-cudf-python`.